src/Npgsql/SingleThreadSynchronizationContext.cs

```diff
@@ -57,17 +63,40 @@ void WorkLoop()
             {
                 var taken = _tasks.TryTake(out var callbackAndState, ThreadStayAliveMs);
                 if (!taken)
-                    return;
-                callbackAndState.Callback(callbackAndState.State);
+                {
+                    lock (_lockObject)
+                    {
+                        if (_tasks.Count == 0)
+                        {
+                            _doingWork = 0;
+                            return;
+                        }
+                    }
+
+                    continue;
+                }
+
+                try
+                {
+                    Debug.Assert(_doingWork == 1);
+                    callbackAndState.Callback(callbackAndState.State);
+                }
+                catch (Exception e)
+                {
+                    Logger.LogError(e, $"Exception caught in {nameof(SingleThreadSynchronizationContext)}");
+                }
             }
         }
         catch (Exception e)
         {
+            // Here we attempt to catch any exception comming from BlockingCollection _tasks
             Logger.LogError(e, $"Exception caught in {nameof(SingleThreadSynchronizationContext)}");
+            _doingWork = 0;
```

Comment:
Wrap this in the lock, this way all accesses to _doingWork are protected and propagation is guaranteed?

Reply:
No reason, really: `BlockingCollection` should only throw if it was disposed, and at that point `SingleThreadSynchronizationContext` is also disposed.

Reply:
I'd still wrap it, for consistency and in case something changes (i.e. an exception can come from somewhere else in the future). But OK for merging as you wish.

Reply:
Done.